src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -52,80 +74,170 @@ private static SnowflakeDateTimeFormat createDateTimeFormatter() {
-    } catch (Exception e) {
-      throw new SFException(
-          e,
-          ErrorCode.INVALID_ROW,
-          input.toString(),
-          "Input column can't be convert to a valid string");
+  /**
+   * Validates and parses JSON array. Non-array types are converted into single-element arrays. All
+   * types in the array tree must be valid variant types, see {@link
+   * DataValidationUtil#isAllowedVariantType}.
+   *
+   * @param input Object to validate
+   * @return JSON array representing the input
+   */
+  static String validateAndParseArray(Object input) {
+    JsonNode jsonNode = validateAndParseVariantAsJsonTree(input, "ARRAY");
+
+    // Non-array values are ingested as single-element arrays, mimicking the Worksheets behavior
+    if (!jsonNode.isArray()) {
+      jsonNode = objectMapper.createArrayNode().add(jsonNode);
     }
 
+    String output = jsonNode.toString();
     // Throw an exception if the size is too large
-    if (output.length() > BYTES_16_MB) {
-      throw new SFException(
-          ErrorCode.INVALID_ROW,
-          input.toString(),
-          String.format("Array too large. length=%d maxLength=%d", output.length(), BYTES_16_MB));
+    int stringLength = output.getBytes(StandardCharsets.UTF_8).length;
+    if (stringLength > MAX_VARIANT_LENGTH) {
+      throw valueFormatNotAllowedException(
+          jsonNode.toString(),
+          "ARRAY",
+          String.format(
+              "Array too large. length=%d maxLength=%d", stringLength, MAX_VARIANT_LENGTH));
     }
     return output;
   }
 
   /**
-   * Expects string JSON or JsonNode
+   * Validates and parses JSON object. Input is rejected if the value does not represent JSON object
+   * (e.g. String '{}' or Map<String, T>). All types in the object tree must be valid variant types,
+   * see {@link DataValidationUtil#isAllowedVariantType}.
    *
-   * @param input Must be valid string JSON or JsonNode
+   * @param input Object to validate
+   * @return JSON object representing the input
    */
   static String validateAndParseObject(Object input) {
-    String output;
-    try {
-      JsonNode node = objectMapper.readTree(input.toString());
-      output = node.toString();
-    } catch (Exception e) {
-      throw new SFException(
-          e, ErrorCode.INVALID_ROW, input.toString(), "Input column can't be convert to Json");
+    JsonNode jsonNode = validateAndParseVariantAsJsonTree(input, "OBJECT");
+    if (!jsonNode.isObject()) {
+      throw valueFormatNotAllowedException(jsonNode, "OBJECT", "Not an object");
     }
 
-    if (output.length() > BYTES_16_MB) {
-      throw new SFException(
-          ErrorCode.INVALID_ROW,
-          input.toString(),
-          String.format("Object too large. length=%d maxLength=%d", output.length(), BYTES_16_MB));
+    String output = jsonNode.toString();
+    // Throw an exception if the size is too large
+    int stringLength = output.getBytes(StandardCharsets.UTF_8).length;
```

Comment:
These checks should happen after parsing because if the user passes a string consisting of millions of whitespace characters and then `{}`, for example, the result would be just `{}`, which would also be ingested. Running the check before would be unnecessarily restrictive.

Reply:
We are essentially checking the length of the minimised JSON, not the pretty-printed JSON passed by the user.